lib/srv/desktop/rdp/rdpclient/client.go

```diff
@@ -412,18 +412,103 @@ func (c *Client) start() {
 					defer C.free(unsafe.Pointer(path))
 					if errCode := C.handle_tdp_sd_info_response(c.rustClient, C.CGOSharedDirectoryInfoResponse{
 						completion_id: C.uint32_t(m.CompletionID),
-						err_code:      C.uint32_t(m.ErrCode),
+						err_code:      m.ErrCode,
 						fso: C.CGOFileSystemObject{
 							last_modified: C.uint64_t(m.Fso.LastModified),
 							size:          C.uint64_t(m.Fso.Size),
-							file_type:     C.uint32_t(m.Fso.FileType),
+							file_type:     m.Fso.FileType,
 							path:          path,
 						},
 					}); errCode != C.ErrCodeSuccess {
 						c.cfg.Log.Errorf("SharedDirectoryInfoResponse failed: %v", errCode)
 						return
 					}
 				}
+			case tdp.SharedDirectoryCreateResponse:
+				if c.cfg.AllowDirectorySharing {
+					if errCode := C.handle_tdp_sd_create_response(c.rustClient, C.CGOSharedDirectoryCreateResponse{
+						completion_id: C.uint32_t(m.CompletionID),
+						err_code:      m.ErrCode,
+					}); errCode != C.ErrCodeSuccess {
+						c.cfg.Log.Errorf("SharedDirectoryCreateResponse failed: %v", errCode)
+						return
+					}
+				}
+			case tdp.SharedDirectoryDeleteResponse:
+				if c.cfg.AllowDirectorySharing {
+					if errCode := C.handle_tdp_sd_delete_response(c.rustClient, C.CGOSharedDirectoryDeleteResponse{
+						completion_id: C.uint32_t(m.CompletionID),
+						err_code:      m.ErrCode,
+					}); errCode != C.ErrCodeSuccess {
+						c.cfg.Log.Errorf("SharedDirectoryDeleteResponse failed: %v", errCode)
+						return
+					}
+				}
+			case tdp.SharedDirectoryListResponse:
+				if c.cfg.AllowDirectorySharing {
+					fsoList := make([]C.CGOFileSystemObject, 0, len(m.FsoList))
+
+					for _, fso := range m.FsoList {
+						path := C.CString(fso.Path)
+						defer C.free(unsafe.Pointer(path))
+
+						fsoList = append(fsoList, C.CGOFileSystemObject{
+							last_modified: C.uint64_t(fso.LastModified),
+							size:          C.uint64_t(fso.Size),
+							file_type:     fso.FileType,
+							path:          path,
+						})
+					}
+
+					fsoListLen := len(fsoList)
+					var cgoFsoList *C.CGOFileSystemObject
+
+					if fsoListLen > 0 {
+						cgoFsoList = (*C.CGOFileSystemObject)(unsafe.Pointer(&fsoList[0]))
+					} else {
+						cgoFsoList = (*C.CGOFileSystemObject)(unsafe.Pointer(&fsoList))
```

Comment:
Should `cgoFsoList` just be nil in this case?

Reply:
No, because it's passed into `handle_tdp_sd_list_response`

https://github.com/gravitational/teleport/blob/5c67a421c01d16ac15bafb45a262f1508cd782f3/lib/srv/desktop/rdp/rdpclient/src/lib.rs#L897-L909

which has the safety requirement that it must be a valid pointer as according to https://doc.rust-lang.org/std/slice/fn.from_raw_parts_mut.html, which says "data must be non-null and aligned even for zero-length slices."

`nil` would (afaict from browsing the internet) be equivalent to passing it a `NULL` pointer. Regardless of whether `nil` is actually address `0` when passed to C, it would not be aligned. OTOH the pointer returned to us by `make` gives us an [aligned](https://go101.org/article/memory-layout.html) [address](https://stackoverflow.com/a/45997533/6277051).